File: analysis/Intensity Matlab Codes/computeBlockFFt.m
function [Xss,numblocks] = computeBlockFFt(data,ns,w,W)
% INPUT:
% data is the time domain data. It should be a single vector.
% ns is the blocksize
% w is the window
% W is the weight (W = mean(W.^2))
% OUTPUT:
% Xss is the single sided scaled data
% numblocks is the number of blocks that the data was broken up into

% Make sure that the is in the right format.
if size(data,1) == 1 && size(data,2) > 1
    data = data';
end

%%% ZERO MEAN
data = data-mean(data);
%%% LENGTH OF DATA
N = length(data);
%%% NUMBER OF BLOCKS
numblocks = floor(2*N/ns-1);
%%% CHUNK DATA
blockData = [reshape(data(1:((numblocks)*ns/2)),[ns/2,(numblocks)])',...
    reshape(data((1:((numblocks)*ns/2))+ns/2),[ns/2,numblocks])'];
blockData = blockData.*repmat(w,[numblocks,1]);
%%% FFT
X = fft(blockData,ns,2);

%%% Scale data so that conj(Xss).*Xss = Gxx
Xss = X(:,1:floor(ns/2))/ns*sqrt(2/W);

end

File: analysis/Intensity Matlab Codes/computeBlockFFT.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun 25 10:59:09 2021

@author: Corey Dobbs
"""

import numpy as np
import scipy as sci

def computeBlockFFT(data, ns, w, W):
    """

    Parameters
    ----------
    data : array
        Single vector of the data to which we will apply the block fft.
        Must be in time domain.
    ns : float
        The number of samples in the data
        The block size.
    w : array
        Window function. Ex: Hanning window
        Used to ensure that data starts and ends at same spot,
        so as to appear periodic
    W : float
        The weight. Ex: W = mean(w w*) 
        (average of window array times conjugate window array)

    Returns
    -------
    Xss : array
        The single-sided scaled, fourier-transformed data
    num_Blocks : float
        The number of blocks into which the data is divided.

    Author: Corey Dobbs
    Transcripted from Matlab code computeBlockFFt.m in 
    BYU_Acoustics/Energy-based_Acoustics/energy-based-acoustics
    in git.physics.byu.edu 
    
    Last Modified: 6/25/21
    """
    
    #Make sure data is right shape
    if np.size(data,0) == 1 and np.size(data,1) > 1:
        np.transpose(data)
    
    #Zero mean
    data = data - np.mean(data)
    
    #Get length of data
    N = len(data)
    #Get number of blocks
    num_Blocks = 2*N//ns - 1
    
    #Chunk data
    blockData_1 = np.transpose(np.reshape(data[0:(num_Blocks*ns//2)],(ns//2, num_Blocks)))
    blockData_2 = np.transpose(np.reshape(data[0 + ns//2:(num_Blocks*ns//2 + ns//2)],(ns//2, num_Blocks)))
    
    blockData = np.append(blockData_1, blockData_2)
    
    blockData = blockData*np.tile(w,(num_Blocks,1))
    
    #FFT
    X = sci.fft(blockData)
    
    #Scale data
    Xss = X[:,0:ns//2]*np.sqrt(2/W) #scipy.fft already divides by ns, but matlab does not
    
    return Xss, num_Blocks